python/python/bystro/supervised_ppca/gf_generative_pt.py

```diff
@@ -0,0 +1,715 @@
+        return covariance
+
+    def _create_prior(self):
+        """
+        This creates the function representing prior on pararmeters
+
+        Parameters
+        ----------
+        log_prior : function
+            The function representing the log density of the prior
+        """
+        pd = self.prior_options
+
+        def log_prior(trainable_variables):
+            W_ = trainable_variables[0]
+            sigma_ = trainable_variables[1]
+            part1 = -1*pd["weight_W"] * torch.mean(torch.square(W_))
+            part2 = Gamma(pd["alpha"], pd["beta"]).log_prob(sigma_)
+            out = torch.mean(part1+part2)
+            return out
+
+        return log_prior
+
+    def _initialize_variables(self, X):
+        """
+        Initializes the variables of the model
+
+        Parameters
+        ----------
+        X : np.array-like,(n_samples,p)
+            The data
+
+        Returns
+        -------
+        W_ : tf.Variable-like,(n_components,p)
+            The loadings of our latent factor model
+
+        sigmal_ : tf.Float
+            The variance of the model
+        """
+        model = dp.PCA(self.n_components)
+        S_hat = model.fit_transform(X)
+        W_init = model.components_.astype(np.float32)
+        W_ = torch.tensor(W_init, requires_grad=True)
+        X_recon = np.dot(S_hat, W_init)
+        diff = np.mean((X - X_recon) ** 2)
+        sinv = softplus_inverse_np(diff * np.ones(1).astype(np.float32))
+        sigmal_ = torch.tensor(sinv, requires_grad=True)
+        return W_, sigmal_
+
+    def _save_losses(self, i, log_likelihood, log_prior, log_posterior):
+        """
+        Saves the values of the losses at each iteration
+
+        Parameters
+        -----------
+        i : int
+            Current training iteration
+
+        losses_likelihood : tf.Float
+            The log likelihood
+
+        losses_prior : tf.Float
+            The log prior
+
+        losses_posterior : tf.Float
+            The log posterior
+        """
+        self.losses_likelihood[i] = log_likelihood.detach().numpy()
+        if torch.is_tensor(log_prior):
+            self.losses_prior[i] = log_prior.detach().numpy()
+        else:
+            self.losses_prior[i] = log_prior
+        self.losses_posterior[i] = log_posterior.detach().numpy()
+
+    def _store_instance_variables(self, trainable_variables):
+        """
+        Saves the learned variables
+
+        Parameters
+        ----------
+        trainable_variables : list
+            List of tensorflow variables saved
+
+        Sets
+        ----
+        W_ : np.array-like,(n_components,p)
+            The loadings
+
+        sigma2_ : float
+            The isotropic variance
+        """
+        self.W_ = trainable_variables[0].detach().numpy()
+        self.sigma2_ = nn.Softplus()(trainable_variables[1]).detach().numpy()
+
+    def _test_inputs(self, X):
+        """
+        Just tests to make sure data is numpy array
```

Comment:
also, if this method doesn't depend on the instance it's better to make it a classmethod or just a pure function.  That way: a) we're not suggesting to the reader that the validation is input specific, and b) we can't possibly make a mistake by giving this function access to data it doesn't need to have (i.e. the instance's attributes)

Reply:
No this will depend on instance. Where this method starts getting customized is when we perform supervision so we want extra conditions, such as having the same number of observations in both X and y